eventbus/eventbus.go

```diff
@@ -0,0 +1,91 @@
+package eventbus
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/lib/pq"
+	"go.uber.org/zap"
+	"sync"
+	"time"
+)
+
+const channelName = "webhookx"
+
+type EventBus struct {
+	ctx      context.Context
+	cancel   context.CancelFunc
+	nodeID   string
+	dsn      string
+	log      *zap.SugaredLogger
+	mux      sync.Mutex
+	handlers map[string][]func(data []byte)
+}
+
+func NewEventBus(nodeID string, dsn string, log *zap.SugaredLogger) *EventBus {
+	ctx, cancel := context.WithCancel(context.Background())
+	bus := EventBus{
+		ctx:      ctx,
+		cancel:   cancel,
+		nodeID:   nodeID,
+		dsn:      dsn,
+		mux:      sync.Mutex{},
+		handlers: make(map[string][]func(data []byte)),
+		log:      log,
+	}
+
+	return &bus
+}
+
+func (bus *EventBus) Start() error {
+	listener := pq.NewListener(bus.dsn, 10*time.Second, time.Minute, nil)
+	err := listener.Listen(channelName)
+	if err != nil {
+		return err
+	}
+	go bus.listenLoop(listener)
+	return nil
+}
+
+func (bus *EventBus) Stop() error {
+	bus.cancel()
+	return nil
+}
+
+func (bus *EventBus) listenLoop(listener *pq.Listener) {
+	defer listener.Close()
+
+	bus.log.Infof("[eventbus] listening on channel: %s", channelName)
+	for {
+		select {
+		case <-bus.ctx.Done():
+			return
+		case n := <-listener.NotificationChannel():
+			var payload EventPayload
+			if err := json.Unmarshal([]byte(n.Extra), &payload); err != nil {
+				bus.log.Errorf("[eventbus] failed to unmarshal payload: %s", err)
+				continue
+			}
+			if payload.Node == bus.nodeID {
+				continue
+			}
+			bus.log.Debugf("[eventbus] received event: channel=%s, payload=%s", n.Channel, n.Extra)
+			if handlers, ok := bus.handlers[payload.Event]; ok {
+				for _, handler := range handlers {
+					handler(payload.Data)
+				}
+			}
+		case <-time.After(90 * time.Second):
```

Comment:
DO NOT USE `time.After` in for select, it will cause memory leak, problem still exist in go 1.22

Reply:
I think `timer.After` does not cause permanent leak, because GC will capture it when the underlying timer expires.

I replaced it by `time.NewTimer()` for the efficiency perspective.